File: examples/example-java/src/main/java/nl/jandt/block/example/PlayerAllowFlight.java
package nl.jandt.block.example;

import net.minestom.server.entity.Player;
import nl.jandt.blocky.engine.core.trait.Behaviour;
import nl.jandt.blocky.engine.impl.trait.PlayerTrait;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class PlayerAllowFlight extends Behaviour {
    private static final Logger log = LoggerFactory.getLogger(PlayerAllowFlight.class);
    private Player player;

    public PlayerAllowFlight() {
        super(false);
    }

    protected void onSetup() {
        parentEventNode().addListener(PlayerTrait.PlayerChangeEvent.class, this::onPlayerChange);

        tryGetTrait(PlayerTrait.class).ifPresent(t -> this.player = t.getPlayer());
    }

    private void onPlayerChange(PlayerTrait.@NotNull PlayerChangeEvent playerChangeEvent) {
        final var oldPlayer = playerChangeEvent.getOldPlayer();
        if (oldPlayer != null) disableFlightForPlayer(oldPlayer);

        this.player = playerChangeEvent.getPlayer();
        enableFlightForPlayer(this.player);
    }

    @Override
    protected void onUpdate() {
//        if (this.player.isFlying())
//            log.trace("Player is flying!");
    }

    @Override
    protected void onEnable() {
        if (this.player == null) return;

        enableFlightForPlayer(this.player);
    }

    @Override
    protected void onDisable() {
        if (this.player == null) return;

        disableFlightForPlayer(this.player);
    }

    public static void enableFlightForPlayer(@NotNull Player player) {
        player.setAllowFlying(true);
        player.setFlying(true);
        log.debug("flight enabled for {}", player);
    }

    public static void disableFlightForPlayer(@NotNull Player player) {
        player.setAllowFlying(false);
        player.setFlying(false);
        log.debug("flight disabled for {}", player);
    }
}

File: examples/example-kotlin/src/main/kotlin/PlayerAllowFlight.kt
package nl.jandt.blocky

import io.github.oshai.kotlinlogging.KotlinLogging
import net.minestom.server.entity.Player
import nl.jandt.blocky.engine.core.trait.Behaviour
import nl.jandt.blocky.engine.impl.trait.PlayerTrait
import nl.jandt.blocky.extensions.parentEventNode

private val log = KotlinLogging.logger {}

class PlayerAllowFlight : Behaviour() {
    private var player: Player? = null

    override fun onSetup() {
        this.isEnabled = false;

        parentEventNode.addListener(PlayerTrait.PlayerChangeEvent::class.java, ::onPlayerChange)

        player = getTrait(PlayerTrait::class.java)?.player
    }

    private fun onPlayerChange(playerChangeEvent: PlayerTrait.PlayerChangeEvent) {
        val oldPlayer = playerChangeEvent.oldPlayer
        if (oldPlayer != null) disableFlightForPlayer(oldPlayer)

        player = playerChangeEvent.player
        enableFlightForPlayer(player!!)
    }

    override fun onEnable() {
        if (this.player == null) return

        enableFlightForPlayer(player!!)
    }

    override fun onDisable() {
        if (this.player == null) return

        disableFlightForPlayer(player!!)
    }

    private fun enableFlightForPlayer(player: Player) {
        player.isAllowFlying = true
        player.isFlying = true
        log.debug { "flight enabled for $player" }
    }

    private fun disableFlightForPlayer(player: Player) {
        player.isAllowFlying = false
        player.isFlying = false
        log.debug { "flight disabled for $player" }
    }
}
